lib/charms/mysql/v0/mysql.py

```diff
@@ -871,12 +874,33 @@ def render_mysqld_configuration(
                 # when memory limit is set, we need to use the minimum
                 # between the available memory and the limit
                 available_memory = min(available_memory, memory_limit)
+
+            if experimental_max_connections:
+                # when set, we use the experimental max connections
+                # and it takes precedence over buffers usage, ensuring minimal memory buffers
+                max_safe_connections = self.get_max_connections(available_memory - MIM_MEM_BUFFERS)
+
+                max_connections = max(
+                    MIN_MAX_CONNECTIONS, min(experimental_max_connections, max_safe_connections)
+                )
+                available_memory -= max_connections * 12 * BYTES_1MiB
+
             (
                 innodb_buffer_pool_size,
                 innodb_buffer_pool_chunk_size,
                 group_replication_message_cache_size,
             ) = self.get_innodb_buffer_pool_parameters(available_memory)
-            max_connections = max(self.get_max_connections(available_memory), MIN_MAX_CONNECTIONS)
+
+            # constrain max_connections based on the available memory
+            # after innodb_buffer_pool_size calculation
+            available_memory -= innodb_buffer_pool_size + (
+                group_replication_message_cache_size or 0
+            )
+            if not experimental_max_connections:
+                max_connections = max(
+                    self.get_max_connections(available_memory), MIN_MAX_CONNECTIONS
```

Comment:
nit (feel free to ignore): `MIN_MAX_CONNECTIONS` is not a minimum, rather a maximum where it is used (if max_safe_connections is < 100, we will bump max_connections up to 100). may be a better name would be `SUGGESTED_MAX_CONNECTIONS` or `MAX_CONNECTIONS_FLOOR` or something like that? 

also out of curiosity, what is the reasoning behind at least 100 max_connections?

Reply:
I'm happy to make it clearer. Right now, the value is guesswork, being smaller than default (151)

Reply:
013a6819